sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -249,13 +255,9 @@ object EnsureRequirements extends Rule[SparkPlan] {
   }
 
   def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
-    // TODO: remove this after we create a physical operator for `RepartitionByExpression`.
-    // SPARK-35989: AQE will change the partition number so we should retain the REPARTITION_BY_NUM
-    // shuffle which is specified by user. And also we can not remove REBALANCE_PARTITIONS_BY_COL,
-    // it is a special shuffle used to rebalance partitions.
-    // So, here we only remove REPARTITION_BY_COL in AQE.
     case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, shuffleOrigin)
-        if shuffleOrigin == REPARTITION_BY_COL || !conf.adaptiveExecutionEnabled =>
+        if optimizeOutRepartition &&
+          (shuffleOrigin == REPARTITION_BY_COL || shuffleOrigin == REPARTITION_BY_NUM) =>
```

Comment:
The logic here is simplified as well. Now it's pretty simple to keep the old logic and try our best to optimize out user-specified repartition.

https://github.com/apache/spark/commit/7fe4c4a9ad92cedd398a13a1781649dff57ca4d9 can be reverted

Reply:
The previous check is to avoid AQE change output partitioning during execution (not the final stage). So I suggested do not remove all shuffle origin before.

Reply:
why do we care about user-specified repartition if it's not in the final stage?

Reply:
User may want to have a custom shuffle partition number with `/*+ repartition(10000, col) */`. If this shuffle is removed, I think it will break user expectaion.

Reply:
If you are referring to something like `df.repartition(1000, col).groupBy(col).agg(....)`, I'm not quite sure if we should keep the user-specified repartition effective, although we do now. cc @maryannxue

Reply:
This may be good or not, but my change here is to revert back to the behavior in 3.1, to make sure we don't have perf regression in 3.2

Reply:
ok I'm fine with this